File: 백준/Silver/1138. 한 줄로 서기/한 줄로 서기.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

// 사람들은 자기보다 큰 사람이 왼쪽에 몇 명 있었는지만을 기억한다.
// N명의 사람이 있고, 사람들의 키는 1부터 N까지 모두 다르다.
const n = Number(input[0])
const height = input[1].split(" ").map(Number)
let str = [n.toString()]
for (let i = n - 2; i >= 0; i--) {
  if(height[i] == 0){
    str = [(i+1).toString(), ...str]
  }else{
    str = [...str.slice(0,height[i]), (i+1).toString(), ...str.slice(height[i])]
  }
}

console.log(str.join(' '))

File: 백준/Silver/1138. 한 줄로 서기/한 줄로 서기.py
import sys
input = sys.stdin.readline

n = int(input())
height = list(map(int,input().split()))

answer =[str(n)]

for i in range(n-2,-1,-1):
  if(height[i] ==0):
    answer = [str(i+1)] +answer
  else:
    answer = answer[0:height[i]]+[str(i+1)]+answer[height[i]:]
print(' '.join(answer))